Sidenav based on screen resize

andOrlando opened the pull request:
Closes #152 
Adds a hamburger with different contents based on screensize. Also comes with a couple css changes for dropdown-related things. For more specific details, see commits

~~Importantly, this doesn't work dynamically so should we add tabs or change tab sizes, we're going to reconfigure the exact screensizes in both the css and the javascript~~ fixed that

andOrlando committed:
adds hamburger with stuff inside it

Adds the hamburger and stuff inside of it. Adds media tags to all the tab buttons. Currently the "Current GPA" dropdown only displays in the sidenav menu

andOrlando committed:
fixes only one gpa dropdown showing also some visual changes

just moves the one gpa dropdown in and out of the navbar, removes the gpa dropdown in the navbar, and updates the order of the elements in the .tab

also does some visual changes like the positioning of the dropdown arrow in the sidenav (sorry psvenk for not doing multiple commits lol)

andOrlando committed:
a bunch of css updates

updates the inaccessible color to make it contrast a little less

updates the dropdown items so that they actually fit correctly. There's a small padding issue with this if someone feels like fixing this

makes everything 270px and float right in the sidenav so it can have a consistant animation

makes the activated item thing for the side nav a little different as to better fit with its theme

also, this one is super important, makes it so that whenever you open one dropdown menu it closes all others becuase otherwise the sidenav freaks out when you have multiple open at once, also this is just a nice ux change

andOrlando committed:
updates sidenav css for dropdown menu

makes it look a little better

andOrlando committed:
makes resizing dynamic

removes hardcoded media queries, instead initializes the resize things later in home.js.

Side note: I now love jQuery

andOrlando committed:
removes log statements

*facepalm*

andOrlando committed:
readability, performance and ui improvements

makes it less of a mess in terms of readability, separating ui changes into functions (like switch_hamburger)

improves performance by only updating ui when it changes from one bracket of screensizes to the next

improves css a little for consistancy (in ui not css, the css is a mess) and some nicer colors.

andOrlando committed:
removes more logs

.... I have nothing to say

andOrlando commented:
~~merged the branch from info-tab to both debug the dynamic screen resize breakpoints and because now there's actually enough space to fit it in nicely. This will also close #76 partially but people should also update the info tab as they see fit.~~

andOrlando committed:
makes reordering dynamic, resizes tableData dropdown, adds info button in sidenav

psvenk commented:
This pull request currently depends on #157 because it adds the info tab to the hamburger menu. I have added the current contents of #157 to this PR, but this can be replaced with the final info tab (using a rebase) when #157 is done.

andOrlando committed:
fixes bug where it wouldn't update when maximizing/minimizing

Since it only detected changes in state 1 apart from eachother it couldn't detect it from 0 to 2 or 2 to 0, which was an oversight on my part. Fixed now

andOrlando committed:
puts items on the left directly into the sidenav

instead of having dupilcates it hides/shows in the sidenav, it just removes displaces them altogether.

also:
- fixes overcomplicated jquerys
- improves a couple comments
- removes tabs in sidenav
- removes weird unecessary init if tree

andOrlando committed:
fixes code formatting

ugly, but apparently consistant with the codebase. Also changes ==s into ===s

andOrlando committed:
convention changes

changes lets to consts and vars to lets

andOrlando commented:
- move hamburger to left
- fix close button
- remove weird color merge in sidenav

andOrlando committed:
makes wednesday changes & bugfix

- removes "close" text and separates close button from rest of sidenav
- moves hamburger over to the far left
- makes hamburger switch exempt
- fixes bug with moving elements in and out of sidenav

andOrlando committed:
makes right items dynamic too

- Fixes a weird bug where it would reorder the sidenav sometimes
- Makes the right items not duplicated in the sidenav
- Seriously cleans up the code for switch_right_items
- Cleans up the code a little for switch_left_items

tektaxi committed:
added sidenav overlay + click anywhere to exit

andOrlando committed:
style changes and bug fix

- makes it so that the width of the tableData dropdown is 210px normally and 270px in the sidenav
- removes the last-child thing which was a bandaid for a problem that was fixed in an earlier commit
- darkens basically all sidenav colors
- makes x a little smaller
- makes colors variables
- gives x button's background color a variable in the style attribute
- makes sidenav-overlay have a normal pointer

psvenk commented:
@andOrlando Looks good. The only remaining blocker is that there is no visual indication of the currently selected tab (as there is in the status quo). We also thought that it would be nice to have an animation when the sidenav changes size, but this can be addressed in a separate pull request (#197).

andOrlando committed:
fixes bug where it doesn't select the header

It still doesn't work in the sidenav, but I think I prefer it that way

psvenk commented on public/home.html:
```diff
@@ -64,18 +65,29 @@
+  <div id="sidenav" class="sidenav">
+    <!-- switch-exempt makes the switch out of navbar pass over this -->
+    <button onclick="closeSideNav()" class="tablinks switch-exempt" style="margin: 0px;width: 270px; height: 44px; background-color: var(--lightgray1)">
+      <i class="fa fa-times fa-2x" style="position: absolute; top: 18px; left: 18px; font-size: 1.3em; color: #000""></i>
```
Remove the double closing quote - while it may still work in some browsers, I think it's still a syntax error and it confuses some syntax highlighters
```suggestion
      <i class="fa fa-times fa-2x" style="position: absolute; top: 18px; left: 18px; font-size: 1.3em; color: #000"></i>
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
 
+let initialize_resize_hamburger = function() {
+
+  //total width of all items in tab minus hamburger
```
Add spacing and capitalization in comments. In this case it would be the following, but you would need to make the change on all of your comments.
```suggestion
  // Total width of all items in tab minus hamburger
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+
+  const switch_left_items = function() {
+    //checks if left items are in sidenav
+    const in_sidenav = $('.tab .tablinks:not(.tablinks-right, .switch-exempt)').length == 0
```
Triple equals
```suggestion
    const in_sidenav = $('.tab .tablinks:not(.tablinks-right, .switch-exempt)').length === 0
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+  //total width of all items in tab minus hamburger
+  let total_width = -44.25;
+
+  //width of all items to the left (because they get removed last) plus logout button and hamburger widths
```
Don't go over 80 characters per line

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+      const children = $('#sidenav .tablinks-right:not(#logout_button)')
+      for (i=0; i<children.length; i++) {
+        $(".tab").append(children[children.length-1-i])
+      }
```
Adding the elements in reverse can be done in a more intuitive, more idiomatic, and less error-prone way than a C-style `for` loop:
```suggestion
      for (const child of [...children].reverse()) {
        $(".tab").append(child);
      }
```
Here, `[...children]` converts `children` to an array. One can then iterate over it in reverse order using a `for`…`of` loop and `Array.reverse()`.

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+      const children = $('.tab .tablinks-right:not(#logout_button)')
+      for (i=0; i<children.length; i++) {
+        $("#sidenav").append(children[children.length-1-i])
+      }
```
As before, this can be done better with a ```for```…```of``` loop:
```suggestion
      for (const child of [...children].reverse()) {
        $("#sidenav").append(child)
      }
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+      //also closes the sidebar
+      closeSideNav();
+    }
+    else {
```
Put `else` on the same line as the closing brace of the `if`:
```suggestion
    } else {
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
 
 }
 
+let initialize_resize_hamburger = function() {
```
Avoid the `let` … `= function()` pattern for top-level functions (because it leads to inconsistent behavior w.r.t. semicolon placement and hoisting):
```suggestion
function initialize_resize_hamburger() {
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+  let left_width = 44.25 + $('#logout_button').outerWidth()
+
+  //gets all non-tablinks-right elements and adds their width to total_width  and also left_width
+  $('.tab .tablinks:not(.tablinks-right)').outerWidth(function(_i, w) {total_width += w; left_width += w;});
```
Use arrow function syntax and don't make overlong (>80 chars) lines; also, there is no need to give a name to the throwaway argument (`_` is fine):
```suggestion
  $('.tab .tablinks:not(.tablinks-right)').outerWidth((_, w) => {
    total_width += w;
    left_width += w;
  });
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+  $('.tab .tablinks:not(.tablinks-right)').outerWidth(function(_i, w) {total_width += w; left_width += w;});
+
+  //gets all tablinks-right elements and adds their width to total_width
+  $('.tab .tablinks-right').outerWidth(function(_i, w) {total_width += w;})
```
Use arrow function syntax:
```suggestion
  $('.tab .tablinks-right').outerWidth((_, w) => {total_width += w;})
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+  //gets all tablinks-right elements and adds their width to total_width
+  $('.tab .tablinks-right').outerWidth(function(_i, w) {total_width += w;})
+
+  const switch_left_items = function() {
```
Use arrow function syntax for all local functions:
```suggestion
  const switch_left_items = () => {
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+      $('#hamburger_button').removeClass("hide")
+    } else {
+      $('#hamburger_button').addClass("hide")
+    }
```
This is needlessly verbose and can be shortened to this:
```suggestion
    if ($('#hamburger_button').hasClass("hide")) {
      $('#hamburger_button').removeClass("hide")
    } else {
      $('#hamburger_button').addClass("hide")
    }
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+  window.addEventListener("resize", function () {
+
+    //updates navbar_state
+    update_navbar_state()
```
Add a semicolon
```suggestion
    update_navbar_state();
```

psvenk commented on public/js/extraFunctions.js:
```diff
@@ -654,6 +654,143 @@ function tableData_closeAllSelect(elmnt) {
+    update_navbar_state()
+
+    //if they're different, updates the tab and sidenav
+    if (old_navbar_state != navbar_state) {
```
Triple equals
```suggestion
    if (old_navbar_state !== navbar_state) {
```

psvenk commented on public/js/home.js:
```diff
@@ -1130,6 +1132,26 @@ function openTab(evt, tab_name) {
+function openSideNav() {
+    let sidenav = document.getElementById("sidenav");
+    if (sidenav.clientWidth == 270) sidenav.style.width = "0px";
+    else sidenav.style.width = "270px";
```
Triple equals and use ternary operator (instead of if/else without braces)
```suggestion
    sidenav.style.width = sidenav.clientWidth === 270 ? "0px" : "270px";
```

psvenk commented on public/js/home.js:
```diff
@@ -1130,6 +1132,26 @@ function openTab(evt, tab_name) {
+    document.getElementById("sidenav").style.width = "0px";
+    const mainSection = $(".sidenav-overlay");
+    $(mainSection).fadeOut("slow");
+    // mainSection.css("visibility", "hidden");
```
Don't keep dead code in comments; we can see any relevant dead code via git.
```suggestion
```

psvenk reviewed (changes requested):
Here are some code style changes. In general, a couple more things:
- Be more consistent in your usage of single vs. double quotes. The convention for this codebase is to use double quotes (at least in the frontend), so it would be good to go with that.
- It would be good to avoid jQuery for new code, but my view is that this code is already written and so it is not as much of a priority here.

andOrlando committed:
Capitalize comments

andOrlando committed:
Apply suggestions from code review

Co-authored-by: psvenk <45520974+psvenk@users.noreply.github.com>

psvenk committed:
Fix issue in home.js from code review

I accidentally deleted some code during code review; this restores that
code.

andOrlando committed:
does the last suggested change

andOrlando committed:
retrofits sidenav for dark theme

also changes some of the variable names which were unused outside of sidenav to be more sidenav oriented (all the lightgrays were only used in sidenav)
